bayesian/utility/user_utils.py

```diff
@@ -22,42 +26,59 @@ def get_user(uuid_val):
-
-        rdb.session.execute(insert_user_stmt)
-        rdb.session.commit()
-        logger.info(f"User created with id {uuid_val}")
+        if snyk_api_token:
+            insert_user_stmt = insert(UserDetails).values(user_id=user_id, user_source=user_source,
+                                                          snyk_api_token=snyk_api_token,
+                                                          created_date=datetime.datetime.now(),
+                                                          status=UserStatus.REGISTERED.name,
+                                                          registered_date=datetime.datetime.now())
+
+            do_update_stmt = insert_user_stmt.on_conflict_do_update(
+                index_elements=['user_id'], set_=dict(snyk_api_token=snyk_api_token,
+                                                      updated_date=datetime.datetime.now()))
+            rdb.session.execute(do_update_stmt)
+            rdb.session.commit()
+            logger.info("User added with id %s", user_id)
     except SQLAlchemyError as e:
         rdb.session.rollback()
-        logger.exception("Error creating new user")
-        raise UserException("Error creating user") from e
+        logger.exception("Error updating user with id %s", user_id)
+        raise UserException("Error updating user") from e
 
 
-@retry(reraise=True, stop=tenacity.stop_after_attempt(3), wait=tenacity.wait_fixed(1))
-def update_user(user_id, snyk_api_token):
-    """Update User."""
+@retry(reraise=True, stop=tenacity.stop_after_attempt(6),
+       wait=tenacity.wait_exponential(multiplier=2, min=4))
+def is_snyk_token_valid(snyk_api_token):
+    """Validate Snyk API token."""
     try:
-        if snyk_api_token:
-            update_stmt = update(UserDetails).where(UserDetails.user_id == user_id).values(
-                dict(snyk_api_token=snyk_api_token, updated_date=datetime.datetime.now(),
-                     status=UserStatus.REGISTERED.name, registered_date=datetime.datetime.now()))
-
-        rdb.session.execute(update_stmt)
-        rdb.session.commit()
-        logger.info(f"User updated with id {user_id}")
-    except SQLAlchemyError as e:
-        rdb.session.rollback()
-        logger.exception(f"Error updating user with id {user_id}")
-        raise UserException("Error updating user") from e
+        response = requests.post(SNYK_API_TOKEN_VALIDATION_URL, json={'api': snyk_api_token},
+                                 headers={'Content-Type': 'application/json'})
+        if response.status_code == 200:
+            return True
+        else:
+            return False
+    except Exception as e:
+        logger.exception("Encountered exception calling Snyk")
+        raise e
+
+
+def encrypt_api_token(snyk_api_token):
+    """Encryption of Api Token."""
+    cipher = Fernet(ENCRYPTION_KEY_FOR_SNYK_TOKEN.encode())
+    return cipher.encrypt(snyk_api_token.encode())
```

Comment:
instead of using single key to encrypt all Snyk API Keys, I suggest to derive a key from UUID for each API key insertion. Something like below,

```
store(uuid, snyk_api_key):
  uuid_hash = sha256(uuid)
  snyk_api_key_encrypted = AES(data=snyk_api_key, key=uuid + salt)
  snyk_api_key_hash = sha256(snyk_api_key)

read(uuid):
  snyk_api_key = AES_decrypt(data=snyk_api_key_encrypted, key=uuid + salt)
```

This approach introduces the following columns to the user database, 
```
uuid_hash - SHA256 hashed value of original UUID
synk_api_key_hash - SHA256 hashed value of original Synk API key to check existence of key without decryption
snyk_api_key_encrypted - AES encrypted Synk API key based original UUID and salt which can be used to query snyk API
```
Also the encryption key for each Snyk API token is different and based on UUID., this ensures that compromise of single encryption key will not expose the rest of the user data as well.

Reply:
Can you let me know why we need to store the hash values. Hashing is one way and  the value cannot be retrieved. There is also no comparison to be done against this value so why should it be stored.
If you are using a salt dont you need to remember the salt for decryption.

Reply:
> Can you let me know why we need to store the hash values

Hashing the Synk API key is meant for comparison, if you want to check whether the Snyk API key is already exists during a call to create_or_update user. Also this can be used to report how many users are using same Snyk API key.

> If you are using a salt dont you need to remember the salt for decryption

Yes, salt should be a global configuration(as an env variable). So compromising either salt or uuid can't expose Snyk API key from DB dump.

Reply:
I am not convinced that this a use case for hashing. We already store the encrypted form and there is no need to do both. For reporting we can always decrypt and figure out how may users are using the same key.

My feeling is that once the salt is leaked it is only a matter of time before it is decrypted. Salts are generally used to ensure that there is no collision when using hashing.

Reply:
Leave about Hashing, what is the problem in using UUID as an encryption key instead of global encryption key?

Reply:
It just seems like the most obvious key that everybody will use if they want to decrypt.

Reply:
Though UUID will be used as key it won't be stored as a plain text in DB., UUID should be hashed and stored.

```
store(uuid, snyk_api_key):
  uuid_hash = sha256(uuid)
  snyk_api_key_encrypted = AES(data=snyk_api_key, key=uuid)

read(uuid):
  snyk_api_key = AES_decrypt(data=snyk_api_key_encrypted, key=uuid)
```

So even if the DB is compromised, data can't be decrypted unless you know each and every UUID.

Reply:
Whats the problem with storing the key as a secret. Thats exactly what it is for. If the secrets get compromised, then we nothing we add there is safe.

Reply:
I'm not talking about the key compromising scenario. If database compromised, having a single key for all data would expose the complete data. However when using UUID, each API key will be encrypted using unique key and it broadly reduces the attack surface.

Reply:
If the key cannot be compromised how can it be decrypted?  I am not seeing any additional value that this can provide.

Reply:
Encrypted data can be compromised using brute force attack. 
With the existing approach, once the attacker gains the key through brute force, then the entire table will be compromised. However in the proposed solution, brute force should be applied for individual rows and would take time.

Reply:
I understand all of this, for this solution I am not convinced that we need to go through all of this. What we have is good enough for the problem that we have.

Reply:
There is a value in data and that is the prime reason for encrypting it. When we choose to encrypt it, why not a hardened approach?. I'm not convinced in going ahead with the existing approach. Thanks.

Reply:
@prashbnair i dont think we need so much of a sophistication for storing the snyk api key. A simple encryption mechanism should be good enough

Reply:
@yzainee Whatever I'm proposing is as simple as the one which is used by @prashbnair . The only difference is selection of single key vs value specific key(UUID).